### PR TITLE
feat(event_source): Extend CodePipeline Artifact Capabilities

### LINKED ISSUE #5447: Feature request: Add Functionality to allow CodePipeline Event to Manage Output Artifacts
### Use case

`CodePipelineJobEvent` allows us to retrieve input artifacts, but does not have functionality to manage and upload output artifacts.

Since the CodePipeline artifact buckets can be very strict, and the event provides credentials, it makes sense to add methods to interact with, and upload, output artifacts.

### Solution/User Experience

I believe this should be a new method like `event.put_artifact` and whatever methods are required to support this.

The bucket, credentials, and encryption configuration should all be auto-magic.

### Alternative solutions

_No response_

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [x] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by mw-root**
**Issue number:** #5447

### Changes

This change adds the ability to upload output artifacts in a simple way, pulling all possible config from the event itself.

### User experience

```python
import json 

from aws_lambda_powertools import Logger
from aws_lambda_powertools.utilities.data_classes import event_source, CodePipelineJobEvent

logger = Logger()

@event_source(data_class=CodePipelineJobEvent)
def lambda_handler(event: CodePipelineJobEvent, context):
    ...
    output_artifact = do_the_thing()
    event.put_artifact(
        artifact_name="json-artifact",
        body=json.dumps(output_artifact),
        content_type="application/json"
    )
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py [hunk 1]
```diff
@@ -236,7 +236,25 @@ def find_input_artifact(self, artifact_name: str) -> CodePipelineArtifact | None
                 return artifact
         return None
 
-    def get_artifact(self, artifact_name: str, filename: str) -> str | None:
+    def find_output_artifact(self, artifact_name: str) -> CodePipelineArtifact | None:
+        """Find an output artifact by artifact name
```
Comment (mw-root) [inline aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py:240]: I'm flexible on this as this is the same behaviour as `find_input_artifact`.

Ultimately, i think that this should fail if the artifact isn't found, but I opted for consistency.

However: uploading an artifact _will_ fail if the artifact is not found as that seems necessary.

File: aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py [hunk 2]
```diff
@@ -255,10 +274,53 @@ def get_artifact(self, artifact_name: str, filename: str) -> str | None:
...
+        bucket = artifact.location.s3_location.bucket_name
+        key = artifact.location.s3_location.key
+
+        if filename:
+            with tempfile.NamedTemporaryFile() as tmp_file:
+                s3.download_file(bucket, key, tmp_file.name)
+                with zipfile.ZipFile(tmp_file.name, "r") as zip_file:
+                    return zip_file.read(filename).decode("UTF-8")
+
+        return s3.get_object(Bucket=bucket, Key=key)["Body"].read()
```
Comment (mw-root) [inline aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py:287]: I have added an option to get non-zip artifacts.

File: aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py [hunk 3]
```diff
@@ -255,10 +274,53 @@ def get_artifact(self, artifact_name: str, filename: str) -> str | None:
...
+            raise ValueError(f"Artifact not found: {artifact_name}.")
+
+        s3 = self.setup_s3_client()
+        bucket = artifact.location.s3_location.bucket_name
+        key = artifact.location.s3_location.key
+        encryption_key_id = self.data.encryption_key.get_id
+        encryption_key_type = self.data.encryption_key.get_type
+
+        if encryption_key_type == "KMS":
+            encryption_key_type = "aws:kms"
```
Comment (mw-root) [inline aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py]: This is an artifact of different APIs/sources providing/requiring different formats of essentially the same value.

**Comment by mw-root** [comment]
Regarding the above issues from SonarCloud:

These are all mocked methods of the S3 client and don't match the `snake_case` convention because boto3 doesn't ;)

**Comment by leandrodamascena** [comment]
Hi @mw-root! Thanks for sending this PR! This week is super busy, but I'll start reviewing it today, ok?

**Comment by mw-root** [comment]
@leandrodamascena glad to do it and no rush at all, thanks for taking it on!

**Comment by anafalcao** [comment]
Hey @mw-root ! I'll be helping @leandrodamascena to review it. Thank you so much for the PR! Great content

File: docs/utilities/data_classes.md [hunk 4]
```diff
@@ -679,7 +679,12 @@ Data classes and utility functions to help create continuous delivery pipelines
             else:
                 template = event.get_artifact(artifact_name, template_file)
                 # Kick off a stack update or create
-                start_update_or_create(job_id, stack, template)
+                result = start_update_or_create(job_id, stack, template)
+                event.put_artifact(
+                    artifact_name="json-artifact",
```
Comment (anafalcao) [inline docs/utilities/data_classes.md]: Would it make more sense to dynamically call the artifact_name here? Like `event.data.input_artifacts[0].name` or `event.data.output_artifacts[0].name`

**Review by anafalcao — COMMENTED** [review]
My initial tests appear to be successful. @leandrodamascena, could you please conduct an additional review of the user experience before we proceed?

**Comment by anafalcao** [comment]
Hi @mw-root , just ran our CI here, and mypy is complaining about the None type here. I saw that `self.data.encryption_key` can be `CodePipelineEncryptionKey` or `None`. 
Can you please take a look?

```sh
Item "None" of "CodePipelineEncryptionKey | None" has no attribute "get_id" 
[union-attr]
            encryption_key_id = self.data.encryption_key.get_id
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py:313:31: error:
Item "None" of "CodePipelineEncryptionKey | None" has no attribute "get_type" 
[union-attr]
            encryption_key_type = self.data.encryption_key.get_type
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Thanks! Let me know if you need help with this

File: aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py (hunk 2 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py:287]: Hey @mw-root! This looks like a super nice improvement to our codebase. 

I'm just curious if there is any situation where artifact is not a zip file, from my understanding, and of course I might be wrong, it's always a zip file. I'll try to find some documentation about this or even simulate this situation.

But anyway, this code looks great to me.

**Review by leandrodamascena — COMMENTED** [review]
Hey @mw-root! Super nice PR! I just left a comment and please try to address it.

**Review by anafalcao — CHANGES_REQUESTED** [review]
Hey! The PR is looking great, can you please take a look in the mypy error? Thanks!

File: docs/utilities/data_classes.md (hunk 4 above)
Comment (mw-root) [reply docs/utilities/data_classes.md]: That works well if you only have one artifact or if you know which order your artifacts returned are in, but to me it feels like an anti-pattern as your artifacts are defined in CodePipeline stage and are named.

Each stage in CodePipeline can have up to 5 artifacts.

File: aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py (hunk 2 above)
Comment (mw-root) [reply aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py:287]: I think as a general case AWS provided jobs will provide zip artifacts, but it's not a requirement. 
For Lambda jobs or custom stages you can upload whatever you want. The CodePipeline framework only defines the s3 location and provides credentials.

In one of our stages we call a Lambda and it uploads both zip and json artifacts that are used by later stages.

**Comment by mw-root** [comment]
@anafalcao 

Sorry about that, commited a fix for the mypy issues.

Thanks!

File: docs/utilities/data_classes.md (hunk 4 above)
Comment (mw-root) [reply docs/utilities/data_classes.md]: Thinking more about this, it's likely this is a more common convention and aligns more with the docs as a whole. I'll update it.

Comment (mw-root) [reply docs/utilities/data_classes.md]: Pushed an update to the docs.

**Comment by mw-root** [comment]
I'll deal with the coverage gap later today.

**Comment by anafalcao** [comment]
> I'll deal with the coverage gap later today.

Awesome, thank you!

**Comment by mw-root** [comment]
I've pushed a commit that should take care of codecov.

**Comment by mw-root** [comment]
Hi @anafalcao, just checking in.. I _think_ I caught everything up.. can you confirm?

**Comment by anafalcao** [comment]
> Hi @anafalcao, just checking in.. I _think_ I caught everything up.. can you confirm?

Hey @mw-root , sorry for the late reply, it was a long weekend in my country. I'm back today and I'll review it now. Thank you very much for your proactivity in solving this quickly

File: aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py (hunk 2 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/data_classes/code_pipeline_job_event.py:287]: Hey @mw-root! Yeah, thanks for explaining this, make sense! 

Resolving the conversation.

**Comment by mw-root** [comment]
No need to apologize @anafalcao, thanks again and I hope you had a good long weekend!

**Review by anafalcao — APPROVED** [review]
Thank you for addressing all the feedback! You've made great improvements to the code, tests and doc. Well done!!

**Review by leandrodamascena — APPROVED** [review]

**Pull request merged**